src/basic/utf8.c

```diff
@@ -150,16 +150,20 @@ bool utf8_is_printable_newline(const char* str, size_t length, bool allow_newlin
         return true;
 }
 
-char *utf8_is_valid(const char *str) {
-        const char *p;
+char *utf8_is_valid_n(const char *str, size_t len_bytes) {
+        /* Check if the string is composed of valid utf8 characters. If length len_bytes is given, stop after
+         * len_bytes. Otherwise, stop at NUL. */
 
         assert(str);
 
-        p = str;
-        while (*p) {
+        for (const char *p = str; len_bytes != (size_t) -1 ? (size_t) (p - str) < len_bytes : *p != '\0'; ) {
                 int len;
 
-                len = utf8_encoded_valid_unichar(p, (size_t) -1);
+                if (len_bytes != (size_t) -1 && *p == '\0')
```

Comment:
doesn't matter, but I'd reverse the two checks given the second check is gong to hit in exceptions only while the other one will either always hit, or never, and we are in an inner loop here...

Reply:
```diff
--- src/basic/utf8.c
+++ src/basic/utf8.c
@@ -159,12 +159,12 @@ char *utf8_is_valid_n(const char *str, size_t len_bytes) {
         for (const char *p = str; len_bytes != (size_t) -1 ? (size_t) (p - str) < len_bytes : *p != '\0'; ) {
                 int len;
 
-                if (len_bytes != (size_t) -1 && *p == '\0')
+                if (_unlikely_(*p == '\0') && len_bytes != (size_t) -1)
                         return NULL; /* embedded NUL */
 
                 len = utf8_encoded_valid_unichar(p,
                                                  len_bytes != (size_t) -1 ? len_bytes - (p - str) : (size_t) -1);
-                if (len < 0)
+                if (_unlikely_(len < 0))
                         return NULL;
 
                 p += len;
```
?

Reply:
Updated like that.